### PR TITLE
Different supports in component distributions for mixture models

### LINKED ISSUE #1780: [FR] Support for different supports in component distributions for mixture models
# Description
I am submitting a feature request to enhance the capabilities of Numpyro in handling mixture models with component distributions having different supports. Currently, when attempting to construct a mixture model using `numpyro.distributions.MixtureGeneral`, all component distributions are required to have the same support.

In my specific use case, I am implementing a truncated power law mixture model where one of the component distributions utilizes a truncated power law model with a high-mass cutoff represented by the Heaviside step function. The other component distribution involves a Gaussian distribution.

The formulation of the model is as follows:

$$
p(m_1\mid\theta) \propto (1-\lambda)A(\theta)m_1^{-\alpha}\Theta(m_{\text{max}}-m_1) + \lambda B(\theta)\exp\left(-\frac{(m_1-\mu_m)^2}{2\sigma_m^2}\right)
$$

Where $\theta$ represents the parameter vector $\left(\lambda,\alpha,m_{\text{min}},\delta_m,m_{\text{max}},\mu_m,\sigma_m\right)$, $A(\cdot)$ and $B(\cdot)$ are normalization factors.

To implement this model efficiently, it is essential to have the flexibility to use component distributions with varying supports within the same mixture model.

Therefore, I kindly request the implementation of this feature to enable users to construct mixture models with component distributions having different supports. If there is any solution which does not requires new feature, will be great.

### PR DESCRIPTION
**Pull request by Qazalbash**
Fixes #1780.

### CODE REVIEW

File: numpyro/distributions/mixtures.py [hunk 1]
```diff
@@ -145,7 +144,7 @@ def sample_with_intermediates(self, key, sample_shape=()):
     def sample(self, key, sample_shape=()):
         return self.sample_with_intermediates(key=key, sample_shape=sample_shape)[0]
 
-    @validate_sample
+    # @validate_sample
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:147]: I think we need to enable `validate_sample` for all components. One way to achieve this is to enable it in the constructor
```
for d in component_distributions:
  d._validate_args = True
```

File: numpyro/distributions/mixtures.py [hunk 2]
```diff
@@ -320,11 +319,9 @@ def __init__(
         # TODO: It would be good to check that the support of all the component
         # distributions match, but for now we just check the type, since __eq__
         # isn't consistently implemented for all support types.
-        support_type = type(component_distributions[0].support)
-        if any(
-            type(d.support) is not support_type for d in component_distributions[1:]
-        ):
-            raise ValueError("All component distributions must have the same support.")
+        # support_type = type(component_distributions[0].support)
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:322]: You can introduce a `support` argument:
```
def __init__(
        self, mixing_distribution, component_distributions, *, support=None, validate_args=None
    ):
```
and we will execute this check when `support=None`.

File: numpyro/distributions/mixtures.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/distributions/mixtures.py:147]: Oops, sorry. You already addressed this in `component_log_probs`.

File: numpyro/distributions/mixtures.py [hunk 3]
```diff
@@ -389,9 +386,21 @@ def component_sample(self, key, sample_shape=()):
...
-            [d.log_prob(value) for d in self.component_distributions], axis=-1
-        )
+        if self._validate_args:
+            mask = jnp.stack(
+                [d.support(value) for d in self.component_distributions], axis=-1
+            )
+            component_log_probs = jnp.where(
+                mask,
+                jnp.stack(
+                    [d.log_prob(value) for d in self.component_distributions], axis=-1
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:396]: It is a little bit cheaper to avoid the duplication of the masking logic
```
component_log_probs = []
for d in self.component_distributions:
  log_prob = d.log_prob(value)
  if (self._support is not None) and (not d._validate_args):
    mask = d.support(value)
    log_prob = jnp.where(mask, log_prob, -jnp.inf)
  component_log_probs.append(log_prob)
...
```

**Review by fehiepsi — COMMENTED** [review]
Thanks for your contribution, @Qazalbash! Could you add a simple test to check whether a mixture of Normal and HalfNormal is computed correctly?

File: numpyro/distributions/mixtures.py (hunk 2 above)
Comment (Qazalbash) [reply numpyro/distributions/mixtures.py:322]: I might be missing something, but I think this check is unnecessary. Can you help me understand why we need to check the types again, even after we've already resolved the issue for different support scenarios?

File: numpyro/distributions/mixtures.py (hunk 3 above)
Comment (Qazalbash) [reply numpyro/distributions/mixtures.py:396]: This makes more sense.

Comment (Qazalbash) [reply numpyro/distributions/mixtures.py:396]: > ```python
> if (self._support is not None) and (not d._validate_args):
> ```
I think `not` is not required with `d._validate_args`.

File: numpyro/distributions/mixtures.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/distributions/mixtures.py:322]: If supports are different, we don't have a mechanism to obtain support of the mixture. We currently use the support of the first component, which only works when all supports are the same.

Comment (Qazalbash) [reply numpyro/distributions/mixtures.py:322]: Seems reasonable, I am adding this change too.

**Comment by Qazalbash** [comment]
@fehiepsi I have made the changes and added the unit tests.

File: numpyro/distributions/mixtures.py [hunk 4]
```diff
@@ -389,9 +397,19 @@ def component_sample(self, key, sample_shape=()):
         return jnp.stack(samples, axis=self.mixture_dim)
 
     def component_log_probs(self, value):
-        component_log_probs = jnp.stack(
-            [d.log_prob(value) for d in self.component_distributions], axis=-1
-        )
+        if self._validate_args:
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:400]: This `if/else` is no longer necessary. The if branch is enough.

File: numpyro/distributions/mixtures.py [hunk 5]
```diff
@@ -389,9 +397,19 @@ def component_sample(self, key, sample_shape=()):
...
+            component_log_probs = []
+            for d in self.component_distributions:
+                log_prob = d.log_prob(value)
+                if (self._support is not None) and d._validate_args:
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:404]: if `d._validate_args` is True, `log_prob` is already masked so we can skip the masking logic here.

File: numpyro/distributions/mixtures.py (hunk 1 above)
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:147]: This is required.

File: numpyro/distributions/mixtures.py [hunk 6]
```diff
@@ -320,11 +324,15 @@ def __init__(
         # TODO: It would be good to check that the support of all the component
         # distributions match, but for now we just check the type, since __eq__
         # isn't consistently implemented for all support types.
-        support_type = type(component_distributions[0].support)
-        if any(
-            type(d.support) is not support_type for d in component_distributions[1:]
-        ):
-            raise ValueError("All component distributions must have the same support.")
+        self._support = support
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:327]: In the property definition `def support(self):`, could you return `self._support` if it is not None? That is the main purpose of this `support` argument (to make sure that MCMC transform is defined correctly)

File: test/test_distributions_mixture.py [hunk 7]
```diff
@@ -78,6 +85,46 @@ def test_mixture_broadcast_batch_shape(
...
+    mixing_probabilities = jnp.ones(2) / 2
+    mixing_distribution = dist.Categorical(probs=mixing_probabilities)
+    component_distribution = [
+        get_normal(batch_shape, validate_args_for_normal),
+        get_half_normal(batch_shape, validate_args_for_half_normal),
+    ]
+    mixture = dist.MixtureGeneral(
+        mixing_distribution=mixing_distribution,
+        component_distributions=component_distribution,
+        support=True,
```
Comment (fehiepsi) [inline test/test_distributions_mixture.py:107]: `support` should be `constraints.real`. Maybe adding a check `is_instance(support, constraints.Constraint` in the mixture constructor to make sure that the argument has correct type.

File: numpyro/distributions/mixtures.py [hunk 8]
```diff
@@ -294,7 +293,12 @@ class MixtureGeneral(_MixtureBase):
     pytree_aux_fields = ("_mixture_size",)
 
     def __init__(
-        self, mixing_distribution, component_distributions, *, validate_args=None
+        self,
+        mixing_distribution,
+        component_distributions,
+        *,
+        support=None,
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:300]: Could you also add docstring to this argument to illustrate the usage?

File: test/test_distributions_mixture.py [hunk 9]
```diff
@@ -78,6 +85,46 @@ def test_mixture_broadcast_batch_shape(
     _test_mixture(mixing_distribution, component_distribution)
 
 
+@pytest.mark.parametrize("batch_shape", [(), (1,), (7,), (2, 5)])
+@pytest.mark.parametrize("validate_args_for_normal", [False, True])
+@pytest.mark.parametrize("validate_args_for_half_normal", [False, True])
```
Comment (fehiepsi) [inline test/test_distributions_mixture.py:90]: Setting `validate_args=True` in the components seems unnecessary to me (assume that `support` is provided like below).

File: numpyro/distributions/mixtures.py (hunk 5 above)
Comment (Qazalbash) [reply numpyro/distributions/mixtures.py:404]: Make sense.

File: test/test_distributions_mixture.py (hunk 7 above)
Comment (Qazalbash) [reply test/test_distributions_mixture.py:107]: I have a question which is not relevant to this PR. I've noticed NumPyro doesn't use type annotations. Is this due to historical reasons, Python's dynamic typing, performance considerations, or building on other frameworks? I'm curious to learn more about the team's approach to code quality and maintainability.

File: numpyro/distributions/mixtures.py [hunk 10]
```diff
@@ -288,13 +291,32 @@ class MixtureGeneral(_MixtureBase):
...
+       >>> component_dists = [
+       ...     dist.Normal(loc=0.0, scale=1.0),
+       ...     dist.HalfNormal(scale=0.3),
+       ... ]
+       >>> mixture = dist.MixtureGeneral(mixing_dist, component_dists, support=dist.constraints.real)
+       >>> mixture.sample(jax.random.PRNGKey(42)).shape
+       ()
     """
 
     pytree_data_fields = ("_mixing_distribution", "_component_distributions")
```
Comment (fehiepsi) [inline numpyro/distributions/mixtures.py:310]: Could you add `_support` to this tuple?  CI is currently failing.

Comment (Qazalbash) [reply numpyro/distributions/mixtures.py]: Done

**Comment by Qazalbash** [comment]
I used python3.11, let me test everything on python3.9!

**Comment by Qazalbash** [comment]
Apparently, all these tests were passing on python3.11. They are still failing on different `atol` values.

File: test/test_distributions_mixture.py [hunk 11]
```diff
@@ -78,6 +85,41 @@ def test_mixture_broadcast_batch_shape(
...
+    mixture = dist.MixtureGeneral(
+        mixing_distribution=mixing_distribution,
+        component_distributions=component_distribution,
+        support=dist.constraints.real,
+    )
+    assert mixture.batch_shape == batch_shape
+    sample_shape = (11,)
+    xx = component_distribution[0].sample(rng_key, sample_shape)
+    log_prob_0 = component_distribution[0].log_prob(xx)
+    log_prob_1 = component_distribution[1].log_prob(xx)
```
Comment (fehiepsi) [inline test/test_distributions_mixture.py:109]: This log prob could be not masked. Maybe this causes tests failing?

Comment (Qazalbash) [reply test/test_distributions_mixture.py:111]: This solved the problem.
